### PR TITLE
Crs 644 overlapping periods validation

### PR DESCRIPTION
**Pull request by ldlharper**
I've also refactored in way we mock the prison-api, to use json files instead. It was getting a bit messy

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/Booking.kt [hunk 1]
```diff
@@ -19,10 +19,10 @@ data class Booking(
...
 
-  fun getOrZero(adjustmentType: AdjustmentType, sentenceAt: LocalDate): Int {
+  fun getOrZero(adjustmentType: AdjustmentType, adjustmentsFrom: LocalDate): Int {
     if (adjustments.containsKey(adjustmentType) && adjustments[adjustmentType] != null) {
       val adjustments = adjustments[adjustmentType]!!
-      val adjustmentDays = adjustments.filter { it.appliesToSentencesFrom.isBeforeOrEqualTo(sentenceAt) }
+      val adjustmentDays = adjustments.filter { it.appliesToSentencesFrom.isBeforeOrEqualTo(adjustmentsFrom) }
         .map { it.numberOfDays }
       if (adjustmentDays.isNotEmpty()) {
         return adjustmentDays.reduce { acc, numberOfDates -> acc + numberOfDates }
```
Comment (yasinmustafa) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/Booking.kt:28]: Is `numberOfDates` named correctly here? is this the totalNumberOfDays?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt [hunk 2]
```diff
@@ -64,19 +66,63 @@ class BookingTimelineService(
...
+      val remandRanges = remandPeriods.map { LocalDateRange.of(it.fromDate, it.toDate) }
+      val sentenceRanges = booking.getAllExtractableSentences().map { LocalDateRange.of(it.sentencedAt, it.sentenceCalculation.adjustedReleaseDate) }
+
+      val allRanges = (remandRanges + sentenceRanges).sortedBy { it.start }
+      var totalRange: LocalDateRange? = null
+      var previousRangeIsRemand: Boolean? = null
+      var previousRange: LocalDateRange? = null
+
+      allRanges.forEach {
+        val isRemand = remandRanges.any() { sentenceRange -> sentenceRange === it }
```
Comment (yasinmustafa) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt:100]: Dont need brackerts here I think `any()`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt [hunk 3]
```diff
@@ -64,19 +66,63 @@ class BookingTimelineService(
...
+          if (it.end.isAfter(totalRange!!.end)) {
+            totalRange = LocalDateRange.of(totalRange!!.start, it.end)
+          }
+          previousRangeIsRemand = isRemand
+          previousRange = it
+        }
+      }
+    }
+  }
+
```
Comment (yasinmustafa) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt:125]: could this whole section be changed to something like this? I think logica,lly it's the same and a bit easier on the eye
```

      allRanges.forEach {
        val isRemand = remandRanges.any { sentenceRange -> sentenceRange === it }
        if (totalRange == null && previousRangeIsRemand == null) {
          totalRange = it
        } else if (it.isConnected(totalRange) && (previousRangeIsRemand!! || isRemand)) {
          // Remand overlaps
          if (previousRangeIsRemand!! && isRemand) {
            throw RemandPeriodOverlapsWithRemandException("Remand of range ${previousRange!!} overlaps with remand of range $it")
          } else {
            throw RemandPeriodOverlapsWithSentenceException("${if (previousRangeIsRemand!!) "Remand" else "Sentence"} of range ${previousRange!!} overlaps with ${if (isRemand) "remand" else "sentence"} of range $it")
          }
        } else if (it.end.isAfter(totalRange!!.end)) {
          totalRange = LocalDateRange.of(totalRange!!.start, it.end)
        }
        previousRangeIsRemand = isRemand
        previousRange = it
      }
```

File: src/test/resources/test_data/api_integration/adjustments/REM-REM.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (yasinmustafa) [inline src/test/resources/test_data/api_integration/adjustments/REM-REM.json:1]: good idea to move this into dedicated json - nice one!

**Review by yasinmustafa — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt (hunk 2 above)
Comment (ldlharper) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt:100]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt (hunk 3 above)
Comment (ldlharper) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt:125]: Yea done, good spot!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/Booking.kt (hunk 1 above)
Comment (ldlharper) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/Booking.kt:28]: I think the acc is the totalNumberOfDays in the reduce function. Thats just how it's named in the kotlin docs.

**Review by yasinmustafa — APPROVED** [review]

**Pull request merged**